### PR TITLE
AB#23 - PR

### PR DESCRIPTION
**Pull request by minhman97**
[AB#23](https://dev.azure.com/minhmancarter/7844c8b2-5ce7-4cec-baf2-5a3248015162/_workitems/edit/23) - PR

### CODE REVIEW

File: .github/workflows/azure-static-web-apps-blue-grass-01e27af00.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    paths:  
+      - "!BackEnd/**"  
+  pull_request:
+    branches: [ "master" ]
+    paths:  
+      - "!BackEnd/**"
+
+jobs:
+  build_and_deploy_job:
+    if: github.event_name == 'push' || (github.event_name == 'pull_request' && github.event.action != 'closed')
```
Comment (minhman97) [inline .github/workflows/azure-static-web-apps-blue-grass-01e27af00.yaml]: remove this

File: .github/workflows/azure-static-web-apps-blue-grass-01e27af00.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        with:
+          submodules: true
+      - name: Build And Deploy
+        id: builddeploy
+        uses: Azure/static-web-apps-deploy@v1
+        with:
+          azure_static_web_apps_api_token: ${{ secrets.AZURE_STATIC_WEB_APPS_API_TOKEN_BLUE_GRASS_01E27AF00 }}
+          repo_token: ${{ secrets.GITHUB_TOKEN }} # Used for Github integrations (i.e. PR comments)
+          action: "upload"
+          ###### Repository/Build Configurations - These values can be configured to match your app requirements. ######
```
Comment (minhman97) [inline .github/workflows/azure-static-web-apps-blue-grass-01e27af00.yaml]: remove comments

File: .github/workflows/dotnet.yml [hunk 3]
```diff
@@ -6,21 +6,68 @@ name: .NET
...
+        
+    - name: Build and push frontend Docker images
+      uses: docker/build-push-action@v4.1.0
+      with:
+        file: FrontEnd/snippet_management.web/Dockerfile
+        context: FrontEnd/snippet_management.web/
+        push: true
+        tags: ${{ secrets.DOCKER_USERNAME }}/snippet-management-web:${{ github.sha }}
+        
+    - name: Azure CLI login and deploy
```
Comment (minhman97) [inline .github/workflows/dotnet.yml:73]: Deploy and config on master push only

File: BackEnd/SnippetManagement.DataModel/Snippet.cs [hunk 4]
```diff
@@ -8,6 +8,16 @@ public class Snippet: BaseEntity<Guid>
     public string Description { get; set; }
     public string Origin { get; set; }
     public string Language { get; set; }
-    public User User { get; set; }
-    public IEnumerable<SnippetTag> Tags { get; set; }
+    public User? User { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.DataModel/Snippet.cs]: user should not nullable

File: BackEnd/SnippetManagement.Service/Model/SnippetTagDto.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@ namespace SnippetManagement.Service.Model;
 public class SnippetTagDto
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Model/SnippetTagDto.cs]: REmove this dto

File: BackEnd/SnippetManagement.Service/Model/SnippetDto.cs [hunk 6]
```diff
@@ -11,5 +24,5 @@ public class SnippetDto
     public DateTimeOffset? Modified { get; set; }
     public string Language { get; set; }
     public Guid UserId { get; set; }
-    public IEnumerable<TagDto> Tags { get; set; }
+    public IEnumerable<TagDto>? Tags { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Model/SnippetDto.cs]: This list should not be nullable

File: BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs [hunk 7]
```diff
@@ -4,24 +4,45 @@ namespace SnippetManagement.Service.Requests;
...
+        Language = language;
+    }
+
     public string Name { get; set; }
     public string Content { get; set; }
     public string Description { get; set; }
     public string Origin { get; set; }
     public string Language { get; set; }
-    public IEnumerable<CreateTagRequest> Tags { get; set; }
+    public IEnumerable<CreateTagRequest>? Tags { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs]: Consider not nullable

File: BackEnd/SnippetManagement.Service/Requests/TagRequest.cs [hunk 8]
```diff
@@ -2,11 +2,11 @@ namespace SnippetManagement.Service.Requests;
...
-    public string TagName { get; set; }
-}
+    public CreateTagRequest(string tagName)
+    {
+        TagName = tagName;
+    }
 
-public class FilterTag
-{
+    public Guid? Id { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/TagRequest.cs]: Check this

File: BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs [hunk 9]
```diff
@@ -12,9 +12,9 @@ namespace SnippetManagement.Service.Services.Implementation;
 
 public class JwtConfiguration
 {
-    public string IssuerSigningKey { get; set; }
-    public string ValidAudience { get; set; }
-    public string ValidIssuer { get; set; }
+    public string? IssuerSigningKey { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs]: These should always be valid

File: BackEnd/SnippetManagement.Test/Api/SnippetTest.cs [hunk 10]
```diff
@@ -38,19 +39,11 @@ public SnippetTest(CustomWebApplicationFactory<Program> factory)
     public async Task CreateSnippet_ShouldBeSuccessful()
     {
         //tạo cai object Snippet
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Test/Api/SnippetTest.cs]: remove comment

File: BackEnd/SnippetManagement.Test/Api/SnippetTest.cs [hunk 11]
```diff
@@ -116,11 +101,11 @@ public async Task UpdateSnippet_ShouldBeSuccessful()
             .FirstOrDefaultAsync(x => x.Id == new Guid($"07785b4a-04e6-4435-b156-63fce124b314"));
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Test/Api/SnippetTest.cs]: should assume this always returns value

File: FrontEnd/snippet_management.web/src/App.js [hunk 12]
```diff
@@ -19,7 +19,7 @@ function App() {
   
   const queryClient = new QueryClient();
   queryClient.invalidateQueries({ queryKey: ["list-snippet"] });
-
+  console.log(token);
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/App.js]: rm this

File: .gitignore [hunk 13]
```diff
@@ -399,3 +399,10 @@ FodyWeavers.xsd
 .idea/
 .vscode/
 
+/BackEnd/SnippetManagement.Api.sln.DotSettings
+
+.env
```
Comment (minhman97) [inline .gitignore]: use pattern

File: BackEnd/SnippetManagement.Api/Program.cs [hunk 14]
```diff
@@ -33,37 +36,30 @@
     {
         ValidateIssuerSigningKey = true,
         IssuerSigningKey =
-            new SymmetricSecurityKey(Encoding.UTF8.GetBytes(builder.Configuration["Jwt:IssuerSigningKey"])),
+            new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwtSection["IssuerSigningKey"] ?? throw new InvalidOperationException())),
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Program.cs]: Validate config on start

File: BackEnd/SnippetManagement.Api/Program.cs [hunk 15]
```diff
@@ -33,37 +36,30 @@
...
 builder.Services.Configure<JwtConfiguration>(options => builder.Configuration.GetSection("Jwt").Bind(options));
 
-Console.WriteLine("checking URL Port: " + builder.Configuration["WEB_PORT"]);
 builder.Services.AddCors(opts =>
 {
     opts.AddPolicy(name: "_myAllowSpecificOrigins",
         policy =>
         {
-            policy.WithOrigins($"http://localhost:{builder.Configuration["Ports:ReactAppPort"]}")
+            policy.WithOrigins(builder.Configuration.GetSection("Cors")["AllowOrigins"]?.Split(",") ?? throw new InvalidOperationException())
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Program.cs]: Similarly

File: BackEnd/SnippetManagement.DataModel/Tag.cs [hunk 16]
```diff
@@ -4,5 +4,10 @@ public class Tag: BaseEntity<Guid>
 {
     public string TagName { get; set; }
 
-    public IEnumerable<SnippetTag> Snippets { get; set; }
+    public IEnumerable<SnippetTag>? Snippets { get; set; }
```
Comment (minhman97) [inline BackEnd/SnippetManagement.DataModel/Tag.cs]: No need for null

File: BackEnd/SnippetManagement.Service/Repositories/ISnippetRepository.cs [hunk 17]
```diff
@@ -7,7 +7,7 @@ namespace SnippetManagement.Service.Repositories;
 
 public interface ISnippetRepository : IRepository<Snippet>
 {
-    Task<Snippet?> Find(Guid id);
+    new Task<Snippet?> Find(Guid id);
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/ISnippetRepository.cs]: Check this

File: BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs [hunk 18]
```diff
@@ -166,20 +159,11 @@ public async Task<PagedResponse<IEnumerable<SnippetDto>>> Search(SearchSnippetRe
             query = query.Where(x => x.Created <= request.ToDate.Value.AddDays(1));
         var totalRecords = await query.CountAsync();
         var snippets = await query.Skip((request.PageNumber - 1) * request.PageSize).Take(request.PageSize)
-            .AsNoTracking().Select(snippet => new SnippetDto()
+            .AsNoTracking().Select(snippet => new SnippetDto(snippet.Id, snippet.Name, snippet.Content,
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs]: move to appropriate place

File: BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs [hunk 19]
```diff
@@ -15,13 +24,25 @@ public class CreateSnippetRequest
 public class UpdateSnippetRequest : CreateSnippetRequest
 {
     public Guid Id { get; set; }
+
+    public UpdateSnippetRequest(string name, string content, string description, string origin, string language) : base(name, content, description, origin, language)
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs]: Id in contructor

File: BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs [hunk 20]
```diff
@@ -15,13 +24,25 @@ public class CreateSnippetRequest
...
 public class SearchSnippetRequest : Pagination
 {
-    public string KeyWord { get; set; }
+    public string? KeyWord { get; set; }
+
+    public string[] Terms
+    {
+        get
+        {
+            if (KeyWord != null) return KeyWord.ToLower().Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs]: ternary operator

File: BackEnd/SnippetManagement.Service/Requests/UserRequest.cs [hunk 21]
```diff
@@ -4,6 +4,11 @@ namespace SnippetManagement.Service.Requests;
 
 public class CreateUserRequest
 {
+    public CreateUserRequest(string email)
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/UserRequest.cs]: Full data constructor?

File: BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs [hunk 22]
```diff
@@ -29,47 +37,38 @@ public AuthenticationService(IOptions<JwtConfiguration> jwtConfiguration, IUnitO
...
+        var tokenHandler = new JwtSecurityTokenHandler();
+        SecurityToken securityToken = tokenHandler.CreateToken(tokenDescriptor);
+
+        return Result.Ok(tokenHandler.WriteToken(securityToken));
     }
 
     private SecurityTokenDescriptor GetTokenDescriptor(UserDto user)
     {
-        byte[] securityKey = Encoding.UTF8.GetBytes(_jwtConfiguration.IssuerSigningKey);
+        byte[] securityKey = Encoding.UTF8.GetBytes(_jwtConfiguration.IssuerSigningKey ?? throw new InvalidOperationException());
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs]: remove this

File: BackEnd/SnippetManagement.Test/Api/SnippetTest.cs [hunk 23]
```diff
@@ -73,38 +65,30 @@ public async Task GetSnippet_ShouldBeSuccessful()
             "https://localhost:44395/Snippet/07785b4a-04e6-4435-b156-63fce124b314",
             null);
         var result = await GetContext().Set<Snippet>().Include(x => x.Tags)
-            .FirstOrDefaultAsync(x => x.Id == new Guid($"07785b4a-04e6-4435-b156-63fce124b314"));
+            .FirstAsync(x => x.Id == new Guid($"07785b4a-04e6-4435-b156-63fce124b314"));
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Test/Api/SnippetTest.cs]: Random Guid

File: BackEnd/SnippetManagement.Test/Api/SnippetTest.cs [hunk 24]
```diff
@@ -113,14 +97,14 @@ public async Task UpdateSnippet_ShouldBeSuccessful()
             JsonSerializer.Serialize(snippetUpdate));
 
         var result = await GetContext().Set<Snippet>().Include(x => x.Tags)
-            .FirstOrDefaultAsync(x => x.Id == new Guid($"07785b4a-04e6-4435-b156-63fce124b314"));
+            .FirstAsync(x => x.Id == new Guid($"07785b4a-04e6-4435-b156-63fce124b314"));
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Test/Api/SnippetTest.cs]: Fix this

File: FrontEnd/snippet_management.web/src/api/apiEndpoint.js [hunk 25]
```diff
@@ -12,9 +12,7 @@ export const getSnippetUrl = (
   sortProperty,
   orderWay
 ) => {
-  return filterKeyWord.trim() === ""
-    ? `${baseUrl}/Snippet?startIndex=${startIndex}&endIndex=${endIndex}&property=${sortProperty}&orderWay=${orderWay}`
-    : `${baseUrl}/Snippet/Search?keyWord=${filterKeyWord.trim()}&startIndex=${startIndex}&endIndex=${endIndex}&property=${sortProperty}&orderWay=${orderWay}`;
+  return `${baseUrl}/Snippet?${filterKeyWord.trim() === "" ? `?` : `/Search?keyWord=${filterKeyWord.trim()}&` }startIndex=${startIndex}&endIndex=${endIndex}&property=${sortProperty}&orderWay=${orderWay}`;
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/api/apiEndpoint.js]: fix this

File: BackEnd/SnippetManagement.DataModel/Snippet.cs (hunk 4 above)
Comment (minhman97) [reply BackEnd/SnippetManagement.DataModel/Snippet.cs]: fixed but the comparator file doesn't display as expected

File: BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs (hunk 7 above)
Comment (minhman97) [reply BackEnd/SnippetManagement.Service/Requests/SnippetRequest.cs]: fixed but the comparator file doesn't display as expected

File: .github/workflows/azure-static-web-apps.yaml [hunk 26]
```diff
@@ -0,0 +1,66 @@
+name: Azure Static Web Apps CI/CD
+
+on:
+  push:
+    branches:
+      - master
+      - develop
+      - AB#23
```
Comment (minhman97) [inline .github/workflows/azure-static-web-apps.yaml]: remove

File: .github/workflows/dotnet.yml [hunk 27]
```diff
@@ -5,22 +5,139 @@ name: .NET
...
+      
     - name: Test
       run: dotnet test BackEnd/SnippetManagement.Api.sln --configuration Release --no-build --verbosity normal
+
+  deploy_to_staging:
+    runs-on: ubuntu-latest
+    environment: Staging
+    needs: build
+
+#     if: ${{ github.ref == 'refs/heads/develop' }}
```
Comment (minhman97) [inline .github/workflows/dotnet.yml]: no comment

File: .github/workflows/dotnet.yml [hunk 28]
```diff
@@ -5,22 +5,139 @@ name: .NET
...
       uses: actions/setup-dotnet@v3
       with:
         dotnet-version: 7.0.x
     - name: Build
       run: dotnet build BackEnd/SnippetManagement.Api.sln --configuration Release    
+      
     - name: Test
       run: dotnet test BackEnd/SnippetManagement.Api.sln --configuration Release --no-build --verbosity normal
+
+  deploy_to_staging:
```
Comment (minhman97) [inline .github/workflows/dotnet.yml]: Don't do this

File: BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs [hunk 29]
```diff
@@ -11,25 +11,22 @@ namespace SnippetManagement.Service.Repositories.Implementation;
...
+            .Include(x => x.Tags)!
             .ThenInclude(xx => xx.Tag)
             .SingleOrDefaultAsync(x => x.Id == id && !x.Deleted);
     }
 
     public async Task<PagedResponse<IEnumerable<SnippetDto?>>> GetAll(Pagination pagination)
     {
         var snippets = (await _context.Set<Snippet>().Where(x => !x.Deleted)
-                .Include(x => x.Tags)
+                .Include(x => x.Tags)!
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs]: Should remove

File: BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs [hunk 30]
```diff
@@ -89,13 +89,13 @@ public SnippetRepository(SnippetManagementDbContext context) : base(context)
                                                                   || x.Origin.ToLower().Contains(request.KeyWord)
                                                                   || x.Description.ToLower().Contains(request.KeyWord)
                                                                   || x.Content.ToLower().Contains(request.KeyWord)
-                                                                  || x.Tags.Any(tagDto =>
-                                                                      tagDto.Tag.TagName.ToLower()
+                                                                  || x.Tags!.Any(tagDto =>
+                                                                      tagDto.Tag != null && tagDto.Tag.TagName.ToLower()
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs]: Don't do this

File: BackEnd/SnippetManagement.Service/Requests/TagRequest.cs [hunk 31]
```diff
@@ -2,11 +2,22 @@ namespace SnippetManagement.Service.Requests;
...
+    public CreateTagRequest(string tagName)
+    {
+        TagName = tagName;
+    }
+
     public string TagName { get; set; }
 }
 
-public class FilterTag
+public class TagsExisted
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Requests/TagRequest.cs]: Rename

File: ExternalSocialAuth/credecial_azure.json [hunk 32]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (minhman97) [inline ExternalSocialAuth/credecial_azure.json]: Remove

File: ExternalSocialAuth/credencial_google.json [hunk 33]
```diff
@@ -1 +1,10 @@
-{"web":{"client_id":"448086775090-g62vpn8qbig7gqt2qiin9m/**/6ok5sgislk.apps.googleusercontent.com","project_id":"snippetmanagement-387702","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"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"}}
\ No newline at end of file
+{
```
Comment (minhman97) [inline ExternalSocialAuth/credencial_google.json]: remove

File: BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs [hunk 34]
```diff
@@ -93,9 +93,9 @@ public SnippetRepository(SnippetManagementDbContext context) : base(context)
                                                                       tagDto.Tag.TagName.ToLower()
                                                                           .Contains(request.KeyWord))
                                                                   || x.Created.ToString().Contains(request.KeyWord)
-                                                                  || x.Modified.ToString().Contains(request.KeyWord)));
+                                                                  || x.Modified.ToString()!.Contains(request.KeyWord)));
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs:96]: remove this

File: BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs [hunk 35]
```diff
@@ -152,34 +144,27 @@ public async Task<PagedResponse<IEnumerable<SnippetDto>>> Search(SearchSnippetRe
...
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Repositories/Implementation/SnippetRepository.cs:147]: same

**Review by popemkt — APPROVED** [review]

**Pull request merged**